### PR TITLE
[AD-1056] AWS Signer Integration

### PR DESCRIPTION
**Pull request by affonsov**
### Summary

[AD-1056] AWS Signer Integration

### Description

AWS Signer Integration

### Related Issue

https://bitquill.atlassian.net/browse/AD-1056

### Additional Reviewers
@affonsoBQ
@alinaliBQ
@andiem-bq
@alexey-temnikov
@birschick-bq

### CODE REVIEW

File: .github/workflows/tableau-connector.yml [hunk 1]
```diff
@@ -18,6 +27,72 @@ jobs:
...
+          if [[ $SIGNING_ENABLED = "true" ]]
+          then
+            echo "Signing is enabled. Will attempt to sign"
+            pwd
+            ls -l tableau-connector/target
+          
+            echo "Installing jq"
+            apt-get install jq
+            name=$(ls tableau-connector/target | grep documentdbjdbc.*.taco)
+            version=$(awk -F- '{print $3}' <<<"$name")
```
Comment (birschick-bq) [inline .github/workflows/tableau-connector.yml]: I would say that using `MAJOR_VERSION`, `MINOR_VERSION`, and `PATCH_VERSION` directly would be safer to get the version string. You could export it from the previous step, or just make the previous step code part of the this step.

Comment (affonsov) [reply .github/workflows/tableau-connector.yml]: I didn't find a way to export an environment variable from one step to another step. Do you have any examples?

File: .github/workflows/tableau-connector.yml [hunk 2]
```diff
@@ -18,9 +28,77 @@ jobs:
...
+            ls -l tableau-connector/target
+          
+            echo "Installing jq"
+            sudo apt-get install jq
+            file="./gradle.properties"
+            MAJOR_VERSION=$(grep "MAJOR_VERSION" ${file} | cut -d'=' -f2)
+            MINOR_VERSION=$(grep "MINOR_VERSION" ${file} | cut -d'=' -f2)
+            PATCH_VERSION=$(grep "PATCH_VERSION" ${file} | cut -d'=' -f2)
+          
+            # Upload unsigned .exe to S3 Bucket
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]: Should be .taco

Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            # Upload unsigned .taco to S3 Bucket
```

File: .github/workflows/tableau-connector.yml [hunk 3]
```diff
@@ -18,9 +28,77 @@ jobs:
...
+            echo "Obtaining version id and uploading unsigned .exe to S3 Bucket"
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            echo "Obtaining version id and uploading unsigned .taco to S3 Bucket"
```

File: .github/workflows/tableau-connector.yml [hunk 4]
```diff
@@ -18,9 +28,77 @@ jobs:
...
+               echo "Exiting because unable to retrieve job ID"
+               exit 1
+            fi
+          
+            # Poll signed S3 bucket to see if the signed artifact is there
+            echo "Poll signed S3 bucket to see if the signed artifact is there"
+            aws s3api wait object-exists --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id}
+          
+            # Get signed EXE from S3
+            echo "Get signed EXE from S3"
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            echo "Get signed .taco from S3"
```

File: .github/workflows/tableau-connector.yml [hunk 5]
```diff
@@ -18,9 +28,77 @@ jobs:
         run: ./tableau-connector/build.sh
       - name: View assembled file
         run: ls -l tableau-connector/target
+      - name: Add Version to Taco file
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]: Assuming the Signing Role step is missing - 
```
uses: aws-actions/configure-aws-credentials@v1
      with:
        role-skip-session-tagging: true
        aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
        aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
        aws-region: us-west-2
        role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
        role-external-id: ${{ secrets.AWS_ROLE_EXTERNAL_ID }}
        role-duration-seconds: 3600
```

Comment (affonsov) [reply .github/workflows/tableau-connector.yml]: added

File: .github/workflows/tableau-connector.yml (hunk 1 above)
Comment (birschick-bq) [reply .github/workflows/tableau-connector.yml]: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable

```
echo "version=$MAJOR_VERSION.$MINOR_VERSION.$PATCH_VERSION" >> $GITHUB_ENV
```

```
... ${{env.version}} ...
```

File: .github/workflows/tableau-connector.yml [hunk 6]
```diff
@@ -18,9 +28,86 @@ jobs:
...
+            then
+               echo "Exiting because unable to retrieve job ID"
+               exit 1
+            fi
+          
+            # Poll signed S3 bucket to see if the signed artifact is there
+            echo "Poll signed S3 bucket to see if the signed artifact is there"
+            aws s3api wait object-exists --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id}.taco --debug
+          
+            # Get signed taco from S3
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            # Dowloading signed taco from S3
```

File: .github/workflows/tableau-connector.yml [hunk 7]
```diff
@@ -18,9 +28,86 @@ jobs:
...
+            echo "Get signed .taco from S3"
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            echo "Dowloading signed .taco from S3"
```

File: .github/workflows/tableau-connector.yml [hunk 8]
```diff
@@ -18,9 +28,86 @@ jobs:
...
+            aws s3api get-object --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id} ./tableau-connector/target/documentdbjdbc-${{env.version}}-signed.taco
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            aws s3api get-object --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id}.taco ./tableau-connector/target/documentdbjdbc-${{env.version}}-signed.taco
```

File: .github/workflows/tableau-connector.yml [hunk 9]
```diff
@@ -18,9 +28,86 @@ jobs:
...
+          
+            if [[ $job_id = "" ]]
+            then
+               echo "Exiting because unable to retrieve job ID"
+               exit 1
+            fi
+          
+            # Poll signed S3 bucket to see if the signed artifact is there
+            echo "Poll signed S3 bucket to see if the signed artifact is there"
+            aws s3api wait object-exists --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id}.taco --debug
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]:
```suggestion
            aws s3api wait object-exists --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id}.taco
```

File: .github/workflows/tableau-connector.yml [hunk 10]
```diff
@@ -4,6 +4,16 @@ on:
...
+  workflow_dispatch:
+    inputs:
+      sign:
+        description: Build and Sign Artifacts (for release only)
+        type: boolean
+        required: false
+        default: false
+env:
+  #SIGNING_ENABLED: ${{ github.event.inputs.sign }}
+  SIGNING_ENABLED: true
```
Comment (alexey-temnikov) [inline .github/workflows/tableau-connector.yml]: Before changes are merged this should be reverted.

File: .github/workflows/tableau-connector.yml [hunk 11]
```diff
@@ -18,9 +28,86 @@ jobs:
...
+          role-skip-session-tagging: true
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-west-2
+          role-to-assume: ${{ secrets.AWS_ROLE_TO_ASSUME }}
+          role-external-id: ${{ secrets.AWS_ROLE_EXTERNAL_ID }}
+          role-duration-seconds: 3600
+
+      - name: Sign Executables
+        id: sign-executables-32
```
Comment (birschick-bq) [inline .github/workflows/tableau-connector.yml]:
```suggestion
```

Not needed.

**Review by birschick-bq — COMMENTED** [review]
Can you add this `sign` input to the release procedure in GETTING_STARTED.md?

File: .github/workflows/tableau-connector.yml (hunk 2 above)
Comment (affonsov) [reply .github/workflows/tableau-connector.yml]: fixed

File: .github/workflows/tableau-connector.yml (hunk 11 above)
Comment (affonsov) [reply .github/workflows/tableau-connector.yml]: fixed

File: .github/workflows/gradle.yml [hunk 12]
```diff
@@ -169,3 +174,102 @@ jobs:
...
+
+            if [[ $job_id = "" ]]
+            then
+               echo "Exiting because unable to retrieve job ID"
+               exit 1
+            fi
+
+            # Poll signed S3 bucket to see if the signed artifact is there
+            echo "Poll signed S3 bucket to see if the signed artifact is there"
+            aws s3api wait object-exists --bucket ${{ secrets.AWS_SIGNED_BUCKET }} --key ${{ secrets.AWS_KEY }}-${job_id}.taco
```
Comment (birschick-bq) [inline .github/workflows/gradle.yml]: Shall we also put this in a loop of three tries, like the ODBC signer?

**Review by birschick-bq — APPROVED** [review]
After removing temporary flags for signing.

File: .github/workflows/gradle.yml [hunk 13]
```diff
@@ -169,3 +174,112 @@ jobs:
         path: ~/.gradle/caches
         key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle') }}
         restore-keys: ${{ runner.os }}-gradle
+
+  build-taco:
+    env:
+      #SIGNING_ENABLED: ${{ github.event.inputs.signTacoFile }}
```
Comment (birschick-bq) [inline .github/workflows/gradle.yml]:
```suggestion
      SIGNING_ENABLED: ${{ github.event.inputs.signTacoFile }}
```

File: .github/workflows/gradle.yml [hunk 14]
```diff
@@ -169,3 +174,112 @@ jobs:
...
+      SIGNING_ENABLED: true
```
Comment (birschick-bq) [inline .github/workflows/gradle.yml]:
```suggestion
```

File: .github/workflows/tableau-connector.yml (hunk 10 above)
Comment (affonsov) [reply .github/workflows/tableau-connector.yml]: removed

File: .github/workflows/gradle.yml (hunk 13 above)
Comment (affonsov) [reply .github/workflows/gradle.yml]: removed

File: .github/workflows/gradle.yml (hunk 14 above)
Comment (affonsov) [reply .github/workflows/gradle.yml]: removed

**Pull request merged**